crates/bevy_ptr/src/lib.rs

```diff
@@ -290,3 +290,95 @@ impl<'a, T> UnsafeCellDeref<'a, T> for &'a UnsafeCell<T> {
+/// it *can* be cloned, so the user is responsible for upholding Rust's aliasing rules.
+/// - A `&T` reference may be released to safe code and may co-exist with other `&T` references,
+/// but not with a `&mut T` reference.
+/// - A `&mut T` reference may be released to safe code provided no other `&mut T` or `&T` exists. A `&mut T` must always be unique.
+///
+#[doc(hidden)]
+pub enum SemiSafeCell<'a, T> {
+    /// Cannot produce references.
+    Empty,
+    /// Can produce shared references.
+    Ref(&'a T),
+    /// Can produce one shared or mutable reference before converting to
+    /// [`Ref`](SemiSafeCell::Ref) or [`Empty`](SemiSafeCell::Empty), respectively.
+    Mut(&'a UnsafeCell<T>),
+}
+
+impl<'a, T> SemiSafeCell<'a, T> {
+    /// Constructs a cell from a shared reference.
+    pub fn from_ref(value: &'a T) -> Self {
+        Self::Ref(value)
+    }
+
+    /// Constructs a cell from a mutable reference.
+    pub fn from_mut(value: &'a mut T) -> Self {
+        // SAFETY: `&mut` ensures unique access.
+        unsafe { Self::Mut(&*(value as *mut T as *const UnsafeCell<T>)) }
+    }
+
+    /// Returns a shared reference to the underlying data.
+    ///
+    /// Note: This converts a [`Mut`](SemiSafeCell::Mut) cell to a [`Ref`](SemiSafeCell::Ref) cell afterwards.
+    ///
+    /// # Safety
+    ///
+    /// Caller must ensure there are no active mutable references to the underlying data.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the cell is empty.
+    pub unsafe fn take_ref(&mut self) -> &'a T {
+        match self {
+            Self::Empty => panic!("cell is empty"),
+            Self::Ref(borrow) => borrow,
+            Self::Mut(cell) => {
+                let borrow = &*cell.get();
+                *self = SemiSafeCell::Ref(borrow);
+                borrow
+            }
+        }
+    }
+
+    /// Returns mutable reference to the underlying data.
+    ///
+    /// Note: This converts a [`Mut`](SemiSafeCell::Mut) cell to an [`Empty`](SemiSafeCell::Empty) cell afterwards.
+    ///
+    /// # Safety
+    ///
+    /// Caller must ensure access to the underlying data is unique (no active references, mutable or not).
+    ///
+    /// # Panics
+    ///
+    /// Panics if the cell is empty or was constructed from a shared reference.
+    pub unsafe fn take_mut(&mut self) -> &'a mut T {
+        match self {
+            Self::Empty => panic!("cell is empty"),
+            Self::Ref(_) => panic!("cannot return mutable reference from Ref cell"),
+            Self::Mut(cell) => {
+                let borrow = &mut *cell.get();
+                *self = SemiSafeCell::Empty;
+                borrow
+            }
+        }
+    }
+}
+
+impl<T> Copy for SemiSafeCell<'_, T> {}
+impl<T> Clone for SemiSafeCell<'_, T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+// SAFETY: Executor guarantees that systems with conflicting access do not run at the same time.
```

Comment:
Is this impl correct to make here? SemiSafeCell is a generic that might be used for stuff other than passing around `World` in the scheduler. IMO this impl should be applied to `SemiSafeCell<World>` inside bevy_ecs, or maybe even to a particular wrapper type for `SemiSafeCell<World>` in bevy_ecs.

Reply:
I assumed that if `T` is `Sync`, this cell can be `Send` since it only holds references. Is that wrong? I can do the specific impl or change the safety comment.

The second part of your comment wasn't clear to me, though. Are you suggesting I move the `SemiSafeCell` struct definition back to `bevy_ecs`?

Reply:
My largest issue is the safety comment referencing the executor and the case of `World` in particular, when `SemiSafeCell` could be getting used in other contexts. If there's a justification for this impl entirely separate to the executor I'd be OK with it, but if the justification of the impl is referencing the executor etc. it should be in bevy_ecs and specific to `World`. If not, the safety comment should be changed.

Reply:
This unsafe trait impl no longer exists.